src/libraries/NormalizedAmount.sol

```diff
@@ -1,58 +1,137 @@
 // SPDX-License-Identifier: Apache 2
 pragma solidity >=0.8.8 <0.9.0;
 
-type NormalizedAmount is uint64;
+struct NormalizedAmount {
+    uint64 amount;
+    uint8 decimals;
+}
 
-using {gt as >, lt as <, sub as -, add as +, div, min, unwrap} for NormalizedAmount global;
+library NormalizedAmountLib {
+    uint8 constant NORMALIZED_DECIMALS = 8;
 
-function gt(NormalizedAmount a, NormalizedAmount b) pure returns (bool) {
-    return NormalizedAmount.unwrap(a) > NormalizedAmount.unwrap(b);
-}
+    error AmountTooLarge(uint256 amount);
+    error NumberOfDecimalsNotEqual(uint8 decimals, uint8 decimalsOther);
+    error AmountUnderflows(uint64 amountA, uint64 amountB);
 
-function lt(NormalizedAmount a, NormalizedAmount b) pure returns (bool) {
-    return NormalizedAmount.unwrap(a) < NormalizedAmount.unwrap(b);
-}
+    function wrap(uint64 amt, uint8 dec) internal pure returns (NormalizedAmount memory) {
+        return NormalizedAmount(amt, dec);
+    }
 
-function sub(NormalizedAmount a, NormalizedAmount b) pure returns (NormalizedAmount) {
-    return NormalizedAmount.wrap(NormalizedAmount.unwrap(a) - NormalizedAmount.unwrap(b));
-}
+    function unwrap(NormalizedAmount memory a) internal pure returns (uint64, uint8) {
+        return (a.amount, a.decimals);
+    }
 
-function add(NormalizedAmount a, NormalizedAmount b) pure returns (NormalizedAmount) {
-    return NormalizedAmount.wrap(NormalizedAmount.unwrap(a) + NormalizedAmount.unwrap(b));
-}
+    function getAmount(NormalizedAmount memory a) internal pure returns (uint64) {
+        return a.amount;
+    }
 
-function div(NormalizedAmount a, uint64 b) pure returns (NormalizedAmount) {
-    return NormalizedAmount.wrap(NormalizedAmount.unwrap(a) / b);
-}
+    function getDecimals(NormalizedAmount memory a) internal pure returns (uint8) {
+        return a.decimals;
+    }
 
-function min(NormalizedAmount a, NormalizedAmount b) pure returns (NormalizedAmount) {
-    return a > b ? b : a;
-}
+    function gt(
+        NormalizedAmount memory a,
+        NormalizedAmount memory b
+    ) internal pure returns (bool) {
+        return a.amount > b.amount;
+    }
 
-function unwrap(NormalizedAmount a) pure returns (uint64) {
-    return NormalizedAmount.unwrap(a);
-}
+    function lt(
+        NormalizedAmount memory a,
+        NormalizedAmount memory b
+    ) internal pure returns (bool) {
+        return a.amount < b.amount;
+    }
 
-library NormalizedAmountLib {
-    error AmountTooLarge(uint256 amount);
+    function isZero(NormalizedAmount memory a) internal pure returns (bool) {
+        return (a.amount == 0 && a.decimals == 0);
+    }
 
-    function normalize(uint256 amount, uint8 decimals) internal pure returns (NormalizedAmount) {
-        if (decimals > 8) {
-            amount /= 10 ** (decimals - 8);
+    function sub(
+        NormalizedAmount memory a,
+        NormalizedAmount memory b
+    ) internal pure returns (NormalizedAmount memory) {
+        if (a.amount < b.amount) {
```

Comment:
this is redundant, solidity inserts underflow checks so now we're doing it twice